File: InMoto/Services/RoadbookParser.swift
import Foundation

/// Interpreta un roadbook testuale e lo trasforma in un TripPlan.
///
/// Formato riconosciuto (tollerante all'ordine delle righe dentro le sezioni):
///
///     NOME DEL VIAGGIO
///     Partenza - Arrivo
///     <link Google Maps percorso intero>
///     8h 10' 456Km
///
///     TAPPA 1
///     Autogrill Carcare - Colle della Maddalena
///     (nota libera, es. benzina)
///     <link Google Maps del percorso della tappa>
///     2h 15' 148Km
///     7:30 - 9:45
///
///     PAUSA 1
///     9:45 - 10:30
///     Colle della Maddalena
///     <link Google Maps del luogo>
///
///     ARRIVO
///     18:15
///     <link>
///     Borgo San Dalmazzo
///
///     PER TORNARE CONSIDERA ANCORA:
///     (testo libero → nota finale)
enum RoadbookParser {

    // MARK: - Sezione grezza

    private enum SectionKind {
        case header, tappa(Int), pausa(Int), arrivo, finale
    }

    private struct RawSection {
        let kind: SectionKind
        var lines: [String] = []
    }

    // MARK: - Parse

    /// Analizza il testo. Risolve i link abbreviati via rete (maps.app.goo.gl).
    static func parse(_ text: String) async -> TripPlan? {
        let sections = splitSections(text)
        guard sections.contains(where: { if case .tappa = $0.kind { return true }; return false }) else {
            return nil
        }

        var nome = "Viaggio"
        var sottotitolo = ""
        var totaleKm = 0
        var totaleMin = 0
        var notaFinale = ""
        var items: [TripPlanItem] = []

        for section in sections {
            let f = await interpret(section.lines)

            switch section.kind {
            case .header:
                // prima riga di testo = nome, seconda = sottotitolo
                if let first = f.textLines.first { nome = first }
                if f.textLines.count > 1 { sottotitolo = f.textLines[1] }
                totaleKm = f.km
                totaleMin = f.durataMin

            case .tappa(let n):
                var tappeNomi = f.waypoints.map { $0.displayName }
                var gwps = f.waypoints.map { geocodeInput($0) }
                // Tappa con un solo punto (link a luogo): rendila navigabile
                // partendo dalla fine della tappa precedente
                if gwps.count == 1,
                   let prev = items.last(where: { $0.kind == .tappa }),
                   let prevLastName = prev.tappeNomi.last,
                   let prevLastWp = prev.waypointsGmaps.last {
                    tappeNomi.insert(prevLastName, at: 0)
                    gwps.insert(prevLastWp, at: 0)
                }
                items.append(TripPlanItem(
                    id: "tappa\(n)",
                    kind: .tappa,
                    titolo: "Tappa \(n)",
                    nome: f.textLines.first ?? defaultName(tappeNomi),
                    nota: f.note,
                    oraInizio: f.oraInizio, oraFine: f.oraFine,
                    km: f.km, durataMin: f.durataMin,
                    tappeNomi: tappeNomi,
                    waypointsGmaps: gwps
                ))

            case .pausa(let n):
                items.append(TripPlanItem(
                    id: "pausa\(n)",
                    kind: .pausa,
                    titolo: "Pausa \(n)",
                    nome: f.textLines.first ?? f.waypoints.first?.displayName ?? "Pausa",
                    nota: f.note,
                    oraInizio: f.oraInizio, oraFine: f.oraFine,
                    km: 0, durataMin: 0,
                    tappeNomi: f.waypoints.map { $0.displayName },
                    waypointsGmaps: f.waypoints.map { geocodeInput($0) }
                ))

            case .arrivo:
                items.append(TripPlanItem(
                    id: "arrivo",
                    kind: .arrivo,
                    titolo: "Arrivo",
                    nome: f.textLines.first ?? f.waypoints.first?.displayName ?? "Arrivo",
                    nota: f.note,
                    oraInizio: f.oraInizio, oraFine: f.oraFine,
                    km: 0, durataMin: 0,
                    tappeNomi: f.waypoints.map { $0.displayName },
                    waypointsGmaps: f.waypoints.map { geocodeInput($0) }
                ))

            case .finale:
                let body = f.allText.joined(separator: "\n")
                notaFinale = notaFinale.isEmpty ? body : notaFinale + "\n" + body
            }
        }

        return TripPlan(
            id: UUID().uuidString,
            nome: nome,
            sottotitolo: sottotitolo,
            totaleKm: totaleKm,
            totaleMin: totaleMin,
            notaFinale: notaFinale,
            items: items
        )
    }

    // MARK: - Divisione in sezioni

    private static func splitSections(_ text: String) -> [RawSection] {
        var sections: [RawSection] = [RawSection(kind: .header)]

        for raw in text.components(separatedBy: .newlines) {
            let line = raw.trimmingCharacters(in: .whitespaces)
            guard !line.isEmpty else { continue }
            let upper = line.uppercased()

            if let n = sectionNumber(upper, prefix: "TAPPA") {
                sections.append(RawSection(kind: .tappa(n)))
            } else if let n = sectionNumber(upper, prefix: "PAUSA") {
                sections.append(RawSection(kind: .pausa(n)))
            } else if upper.hasPrefix("ARRIVO") {
                sections.append(RawSection(kind: .arrivo))
            } else if upper.hasPrefix("PER TORNARE") || upper.hasPrefix("NOTE") {
                sections.append(RawSection(kind: .finale))
                // il resto della riga (dopo i due punti) fa parte della nota
                if let colon = line.firstIndex(of: ":") {
                    let rest = line[line.index(after: colon)...].trimmingCharacters(in: .whitespaces)
                    if !rest.isEmpty { sections[sections.count - 1].lines.append(rest) }
                } else {
                    sections[sections.count - 1].lines.append(line)
                }
            } else {
                sections[sections.count - 1].lines.append(line)
            }
        }
        return sections
    }

    private static func sectionNumber(_ upper: String, prefix: String) -> Int? {
        guard upper.hasPrefix(prefix) else { return nil }
        let rest = upper.dropFirst(prefix.count).trimmingCharacters(in: .whitespaces)
        return Int(rest.prefix(while: { $0.isNumber })) ?? (rest.isEmpty ? 1 : nil)
    }

    // MARK: - Interpretazione delle righe di una sezione

    private struct Fields {
        var waypoints: [ParsedWaypoint] = []
        var textLines: [String] = []   // righe "nome" (testo libero non classificato)
        var note: String = ""
        var oraInizio = ""
        var oraFine = ""
        var km = 0
        var durataMin = 0
        var allText: [String] = []     // tutte le righe non-link (per la nota finale)
    }

    private static func interpret(_ lines: [String]) async -> Fields {
        var f = Fields()

        for line in lines {
            if line.lowercased().hasPrefix("http") {
                let url = GoogleMapsParser.isShortURL(line)
                    ? (await GoogleMapsParser.resolveShortURL(line) ?? line)
                    : line
                let (parsed, _) = GoogleMapsParser.parse(url)
                f.waypoints.append(contentsOf: parsed)
                continue
            }
            f.allText.append(line)

            if let (start, end) = timeRange(line) {
                if f.oraInizio.isEmpty { f.oraInizio = start; f.oraFine = end }
                continue
            }
            if let single = singleTime(line) {
                if f.oraInizio.isEmpty { f.oraInizio = single }
                continue
            }
            if line.hasPrefix("(") {
                let cleaned = line.trimmingCharacters(in: CharacterSet(charactersIn: "()"))
                f.note = f.note.isEmpty ? cleaned : f.note + " " + cleaned
                continue
            }
            if let (km, mins) = metrics(line) {
                if km > 0 { f.km = km }
                if mins > 0 { f.durataMin = mins }
                continue
            }
            f.textLines.append(line)
        }
        return f
    }

    /// "7:30 - 9:45", "7.30-9.45" → ("7:30", "9:45")
    private static func timeRange(_ line: String) -> (String, String)? {
        let pattern = #"^(\d{1,2})[:.](\d{2})\s*[-–—]\s*(\d{1,2})[:.](\d{2})$"#
        guard let m = line.range(of: pattern, options: .regularExpression) else { return nil }
        let parts = String(line[m])
            .replacingOccurrences(of: ".", with: ":")
            .components(separatedBy: CharacterSet(charactersIn: "-–—"))
            .map { $0.trimmingCharacters(in: .whitespaces) }
        guard parts.count == 2 else { return nil }
        return (parts[0], parts[1])
    }

    /// "18:15" → "18:15"
    private static func singleTime(_ line: String) -> String? {
        guard line.range(of: #"^\d{1,2}[:.]\d{2}$"#, options: .regularExpression) != nil else { return nil }
        return line.replacingOccurrences(of: ".", with: ":")
    }

    /// "2h 15' 148Km" → (148, 135) · "3h 149 Km" → (149, 180) · "101Km" → (101, 0)
    private static func metrics(_ line: String) -> (km: Int, durataMin: Int)? {
        var km = 0
        var rest = line

        // Prima i km, rimossi dalla riga: altrimenti in "3h 149 Km" il 149
        // verrebbe scambiato per i minuti della durata
        if let r = rest.range(of: #"(\d+)\s*[Kk][Mm]"#, options: .regularExpression) {
            km = Int(rest[r].prefix(while: { $0.isNumber })) ?? 0
            rest.removeSubrange(r)
        }

        var mins = 0
        if let r = rest.range(of: #"(\d+)\s*[Hh](\s*\d{1,2})?"#, options: .regularExpression) {
            let nums = String(rest[r])
                .components(separatedBy: CharacterSet.decimalDigits.inverted)
                .filter { !$0.isEmpty }
                .compactMap { Int($0) }
            if let h = nums.first {
                mins = h * 60 + (nums.count > 1 ? nums[1] : 0)
            }
        }
        guard km > 0 || mins > 0 else { return nil }
        return (km, mins)
    }

    /// Stringa per il geocoding: non forzare ", Italy" sui nomi estratti
    /// dai link (i viaggi possono sconfinare, es. Francia). Le coordinate
    /// passano invariate.
    private static func geocodeInput(_ wp: ParsedWaypoint) -> String {
        if wp.gmapsWaypoint == "\(wp.displayName), Italy" {
            return wp.displayName
        }
        return wp.gmapsWaypoint
    }

    private static func defaultName(_ tappe: [String]) -> String {
        guard let first = tappe.first, let last = tappe.last, tappe.count >= 2 else {
            return tappe.first ?? "Tappa"
        }
        return "\(short(first)) - \(short(last))"
    }

    private static func short(_ s: String) -> String {
        s.components(separatedBy: ",").first ?? s
    }
}

File: android/app/src/main/java/com/divito/inmoto/data/RoadbookParser.kt
package com.divito.inmoto.data

import com.divito.inmoto.model.TripPlan
import com.divito.inmoto.model.TripPlanItem
import java.util.UUID

/**
 * Interpreta un roadbook testuale e lo trasforma in un TripPlan.
 * Port 1:1 del RoadbookParser di iOS. Vedi il formato nel commento originale:
 * sezioni TAPPA n / PAUSA n / ARRIVO / PER TORNARE, orari "7:30 - 9:45",
 * metriche "2h 15' 148Km", note tra parentesi, link Google Maps.
 */
object RoadbookParser {

    // MARK: - Sezione grezza
    private sealed interface SectionKind {
        data object Header : SectionKind
        data class Tappa(val n: Int) : SectionKind
        data class Pausa(val n: Int) : SectionKind
        data object Arrivo : SectionKind
        data object Finale : SectionKind
    }

    private class RawSection(val kind: SectionKind) {
        val lines = mutableListOf<String>()
    }

    // MARK: - Parse

    /** Analizza il testo. Risolve i link abbreviati via rete (maps.app.goo.gl). */
    suspend fun parse(text: String): TripPlan? {
        val sections = splitSections(text)
        if (sections.none { it.kind is SectionKind.Tappa }) return null

        var nome = "Viaggio"
        var sottotitolo = ""
        var totaleKm = 0
        var totaleMin = 0
        var notaFinale = ""
        val items = mutableListOf<TripPlanItem>()

        for (section in sections) {
            val f = interpret(section.lines)

            when (val kind = section.kind) {
                is SectionKind.Header -> {
                    f.textLines.firstOrNull()?.let { nome = it }
                    if (f.textLines.size > 1) sottotitolo = f.textLines[1]
                    totaleKm = f.km
                    totaleMin = f.durataMin
                }

                is SectionKind.Tappa -> {
                    val nameLine = f.textLines.firstOrNull()
                    val tappeNomi = f.waypoints.map { it.displayName }.toMutableList()
                    val gwps = f.waypoints.map { geocodeInput(it) }.toMutableList()
                    // Tappa con un solo punto: ricava la partenza dalla tappa
                    // precedente, oppure dal nome "Partenza - Arrivo".
                    if (gwps.size == 1) {
                        val prev = items.lastOrNull { it.kind == TripPlanItem.Kind.tappa }
                        val prevLastName = prev?.tappeNomi?.lastOrNull()
                        val prevLastWp = prev?.waypointsGmaps?.lastOrNull()
                        if (prevLastName != null && prevLastWp != null) {
                            tappeNomi.add(0, prevLastName)
                            gwps.add(0, prevLastWp)
                        } else {
                            endpointsFromName(nameLine)?.let { (start, _) ->
                                tappeNomi.add(0, start)
                                gwps.add(0, start)
                            }
                        }
                    } else if (gwps.isEmpty()) {
                        endpointsFromName(nameLine)?.let { (start, end) ->
                            tappeNomi.clear(); tappeNomi.add(start); tappeNomi.add(end)
                            gwps.clear(); gwps.add(start); gwps.add(end)
                        }
                    }
                    items.add(
                        TripPlanItem(
                            id = "tappa${kind.n}",
                            kind = TripPlanItem.Kind.tappa,
                            titolo = "Tappa ${kind.n}",
                            nome = nameLine ?: defaultName(tappeNomi),
                            nota = f.note,
                            oraInizio = f.oraInizio, oraFine = f.oraFine,
                            km = f.km, durataMin = f.durataMin,
                            tappeNomi = tappeNomi,
                            waypointsGmaps = gwps,
                        )
                    )
                }

                is SectionKind.Pausa -> items.add(
                    TripPlanItem(
                        id = "pausa${kind.n}",
                        kind = TripPlanItem.Kind.pausa,
                        titolo = "Pausa ${kind.n}",
                        nome = f.textLines.firstOrNull() ?: f.waypoints.firstOrNull()?.displayName ?: "Pausa",
                        nota = f.note,
                        oraInizio = f.oraInizio, oraFine = f.oraFine,
                        km = 0, durataMin = 0,
                        tappeNomi = f.waypoints.map { it.displayName },
                        waypointsGmaps = f.waypoints.map { geocodeInput(it) },
                    )
                )

                is SectionKind.Arrivo -> items.add(
                    TripPlanItem(
                        id = "arrivo",
                        kind = TripPlanItem.Kind.arrivo,
                        titolo = "Arrivo",
                        nome = f.textLines.firstOrNull() ?: f.waypoints.firstOrNull()?.displayName ?: "Arrivo",
                        nota = f.note,
                        oraInizio = f.oraInizio, oraFine = f.oraFine,
                        km = 0, durataMin = 0,
                        tappeNomi = f.waypoints.map { it.displayName },
                        waypointsGmaps = f.waypoints.map { geocodeInput(it) },
                    )
                )

                is SectionKind.Finale -> {
                    val body = f.allText.joinToString("\n")
                    notaFinale = if (notaFinale.isEmpty()) body else "$notaFinale\n$body"
                }
            }
        }

        return TripPlan(
            id = UUID.randomUUID().toString(),
            nome = nome,
            sottotitolo = sottotitolo,
            totaleKm = totaleKm,
            totaleMin = totaleMin,
            notaFinale = notaFinale,
            items = items,
        )
    }

    // MARK: - Divisione in sezioni

    private fun splitSections(text: String): List<RawSection> {
        val sections = mutableListOf(RawSection(SectionKind.Header))

        for (raw in text.split(Regex("\\r?\\n"))) {
            val line = raw.trim()
            if (line.isEmpty()) continue
            val upper = line.uppercase()

            val tappaN = sectionNumber(upper, "TAPPA")
            val pausaN = sectionNumber(upper, "PAUSA")
            when {
                tappaN != null -> sections.add(RawSection(SectionKind.Tappa(tappaN)))
                pausaN != null -> sections.add(RawSection(SectionKind.Pausa(pausaN)))
                upper.startsWith("ARRIVO") -> sections.add(RawSection(SectionKind.Arrivo))
                upper.startsWith("PER TORNARE") || upper.startsWith("NOTE") -> {
                    sections.add(RawSection(SectionKind.Finale))
                    val colon = line.indexOf(':')
                    if (colon >= 0) {
                        val rest = line.substring(colon + 1).trim()
                        if (rest.isNotEmpty()) sections.last().lines.add(rest)
                    } else {
                        sections.last().lines.add(line)
                    }
                }
                else -> sections.last().lines.add(line)
            }
        }
        return sections
    }

    private fun sectionNumber(upper: String, prefix: String): Int? {
        if (!upper.startsWith(prefix)) return null
        val rest = upper.drop(prefix.length).trim()
        val digits = rest.takeWhile { it.isDigit() }
        return digits.toIntOrNull() ?: if (rest.isEmpty()) 1 else null
    }

    // MARK: - Interpretazione delle righe di una sezione

    private class Fields {
        val waypoints = mutableListOf<ParsedWaypoint>()
        val textLines = mutableListOf<String>()
        var note = ""
        var oraInizio = ""
        var oraFine = ""
        var km = 0
        var durataMin = 0
        val allText = mutableListOf<String>()
    }

    private suspend fun interpret(lines: List<String>): Fields {
        val f = Fields()
        for (line in lines) {
            if (line.lowercase().startsWith("http")) {
                val url = if (GoogleMapsParser.isShortURL(line))
                    (GoogleMapsParser.resolveShortURL(line) ?: line) else line
                val (parsed, _) = GoogleMapsParser.parse(url)
                f.waypoints.addAll(parsed)
                continue
            }
            f.allText.add(line)

            val range = timeRange(line)
            if (range != null) {
                if (f.oraInizio.isEmpty()) { f.oraInizio = range.first; f.oraFine = range.second }
                continue
            }
            val single = singleTime(line)
            if (single != null) {
                if (f.oraInizio.isEmpty()) f.oraInizio = single
                continue
            }
            if (line.startsWith("(")) {
                val cleaned = line.trim('(', ')')
                f.note = if (f.note.isEmpty()) cleaned else "${f.note} $cleaned"
                continue
            }
            val m = metrics(line)
            if (m != null) {
                if (m.first > 0) f.km = m.first
                if (m.second > 0) f.durataMin = m.second
                continue
            }
            f.textLines.add(line)
        }
        return f
    }

    private val timeRangeRegex = Regex("""^(\d{1,2})[:.](\d{2})\s*[-–—]\s*(\d{1,2})[:.](\d{2})$""")
    private val singleTimeRegex = Regex("""^\d{1,2}[:.]\d{2}$""")
    private val kmRegex = Regex("""(\d+)\s*[Kk][Mm]""")
    private val hourRegex = Regex("""(\d+)\s*[Hh](\s*\d{1,2})?""")

    /** "7:30 - 9:45", "7.30-9.45" → ("7:30", "9:45") */
    private fun timeRange(line: String): Pair<String, String>? {
        if (!timeRangeRegex.matches(line)) return null
        val parts = line.replace(".", ":")
            .split(Regex("[-–—]"))
            .map { it.trim() }
        if (parts.size != 2) return null
        return parts[0] to parts[1]
    }

    /** "18:15" → "18:15" */
    private fun singleTime(line: String): String? {
        if (!singleTimeRegex.matches(line)) return null
        return line.replace(".", ":")
    }

    /** "2h 15' 148Km" → (148, 135) · "3h 149 Km" → (149, 180) · "101Km" → (101, 0) */
    private fun metrics(line: String): Pair<Int, Int>? {
        var km = 0
        var rest = line
        kmRegex.find(rest)?.let { m ->
            km = m.value.takeWhile { it.isDigit() }.toIntOrNull()
                ?: m.groupValues[1].toIntOrNull() ?: 0
            rest = rest.removeRange(m.range)
        }
        var mins = 0
        hourRegex.find(rest)?.let { m ->
            val nums = m.value.split(Regex("\\D+")).filter { it.isNotEmpty() }.mapNotNull { it.toIntOrNull() }
            nums.firstOrNull()?.let { h ->
                mins = h * 60 + (if (nums.size > 1) nums[1] else 0)
            }
        }
        if (km <= 0 && mins <= 0) return null
        return km to mins
    }

    /** Non forzare ", Italy" sui nomi estratti dai link (i viaggi sconfinano). */
    private fun geocodeInput(wp: ParsedWaypoint): String =
        if (wp.gmapsWaypoint == "${wp.displayName}, Italy") wp.displayName else wp.gmapsWaypoint

    private val dashRegex = Regex("""\s+[–—]\s+""")

    /** Da "Autogrill Carcare - Colle della Maddalena" ricava (partenza, arrivo). */
    private fun endpointsFromName(name: String?): Pair<String, String>? {
        val raw = name?.trim().orEmpty()
        if (raw.isEmpty()) return null
        val parts = raw.replace(dashRegex, " - ")
            .split(" - ")
            .map { it.trim() }
            .filter { it.isNotEmpty() }
        if (parts.size < 2) return null
        return parts.first() to parts.last()
    }

    /** True se la riga è un metadato (orario/metriche/intestazione), non un luogo. */
    fun isNonPlaceLine(raw: String): Boolean {
        val line = raw.trim()
        if (line.isEmpty()) return true
        val upper = line.uppercase()
        for (h in listOf("TAPPA", "PAUSA", "ARRIVO", "PARTENZA", "PER TORNARE", "NOTE")) {
            if (upper == h || upper.startsWith("$h ")) return true
        }
        if (timeRange(line) != null) return true
        if (singleTime(line) != null) return true
        if (metrics(line) != null) return true
        return false
    }

    private fun defaultName(tappe: List<String>): String {
        val first = tappe.firstOrNull()
        val last = tappe.lastOrNull()
        if (first == null || last == null || tappe.size < 2) return first ?: "Tappa"
        return "${short(first)} - ${short(last)}"
    }

    private fun short(s: String): String = s.split(",").firstOrNull() ?: s
}
